web/src/app/components/bar-chart/bar-chart.component.ts

```diff
@@ -0,0 +1,387 @@
+import { AfterViewInit, Component, ElementRef, Input, ViewChild} from '@angular/core';
+import { formatDate } from '@angular/common'
+import * as d3 from 'd3';
+
+import { step189_2020 } from '../../../proto/step189_2020';
+
+/**
+ * Item includes all data used by the single in the bar chart.
+ */
+interface Item {
+  pushID: string;    // Push ID string
+  state: number;     // Tag of the push end state
+  startTime: string; // Start time of push, in `YYYY-MM-DD HH:mm` format
+  durationHours: number;  // Duration of push, in hours
+}
+
+/**
+ * D3 typs used by the bar chart.
+ * 
+ * The d3.Selection has the defult type Selection<GElement, Datum, PElement,
+ * PDatum>, and we want to use it with Datum, Datum, PElement, PDatum being
+ * `undefined` or `null`.
+ */
+type d3SVG = d3.Selection<SVGSVGElement, undefined, null, undefined>;
+type d3G = d3.Selection<SVGGElement, undefined, null, undefined>;
+type d3ScaleLinear = d3.ScaleLinear<number, number>;
+
+@Component({
+  selector: 'app-bar-chart',
+  templateUrl: './bar-chart.component.html',
+  styleUrls: ['./bar-chart.component.scss']
+})
+
+export class BarChartComponent implements AfterViewInit {
+  /**
+   * Constants.
+   */
+  private static readonly DEFAULT_NUM_BARS: number = 30;
+  private static readonly NANO_TO_SECS: number = 10 ** 9;
+  private static readonly NANO_TO_MILLI: number = 10 ** 6;
+  private static readonly SECS_TO_HRS: number = 60 * 60;
+  private static readonly COLOR_LIGHT_GRAY: string = '#787878';
+  private static readonly COLOR_DARK_GRAY: string = '#373C38';
+  private static readonly COLOR_WHITE: string = '#eee';
+  private static readonly COLOR_WHITE_TRANS: string = '#ffffff00';
+  private static readonly DATE_FORMAT: string = 'yyyy-mm-dd hh:mm';
+  private static readonly DATE_LOCALE: string = 'en-US';
+  private static readonly STATE_TO_COLOR: { [index: number]: string } = {
+    1: '#eee',
+    3: '#2196f3',
+    4: '#d50000',
+    5: '#34a853',
+    6: '#d50000',
+    7: '#2196f3',
+    8: '#2196f3',
+    9: '#d50000',
+    10: '#2196f3',
+    12: '#d50000',
+    13: '#2196f3',
+    15: '#2196f3',
+    16: '#d50000'
+  };
+
+  /**
+   * Private variables.
+   * 
+   * The non-null assertion operator('!') suggests that there is no 
+   * issue of variables being `null` or `undefined`. Some variables are 
+   * declared as any, as their types change in different functions.
+   */
+  @ViewChild('barchart') private barChartContainer!: ElementRef; // Bar chart component
+  @Input() private pushInfos!: step189_2020.IPushInfo[] | null;
+  @Input() private currentPush!: step189_2020.IPushInfo | null;
+
+  private dataAll: Item[] = [];
+  private dataComplete: Item[] = [];
+  private totalDuration: number[] = [];
+  private completeDuration: number[] = [];
+  private svg: d3SVG | undefined;
+  private focus: d3G | undefined;
+  private brush: d3G | undefined;
```

Comment:
Can you add some comments that explain what these are for? Specifically, it would be good to clearly state that focus is top plot and brush is the one from the bottom.